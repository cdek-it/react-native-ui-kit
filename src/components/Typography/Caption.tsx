import type { Icon } from '@tabler/icons-react-native'
import { Text, type TextProps, View } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

export interface CaptionProps extends TextProps {
  readonly color?: 'default' | 'secondary' | 'primary'
  readonly disabled?: boolean
  /** Иконка из набора Tabler */
  readonly Icon?: Icon
}

export const Caption = ({
  color = 'default',
  disabled,
  style,
  Icon,
  ...other
}: CaptionProps) => {
  const styles = useStyles()
  const text = (
    <Text
      style={[
        styles.text,
        styles[color],
        disabled && styles.disabled,
        Icon && styles.textWithIcon,
        style,
      ]}
      {...other}
    />
  )

  if (Icon) {
    return (
      <View style={styles.withIconContainer}>
        <Icon {...styles[color]} {...styles.icon} />
        {text}
      </View>
    )
  }

  return text
}

const useStyles = makeStyles(({ theme, spacing, typography }) => ({
  text: { fontSize: 12.25, includeFontPadding: false, verticalAlign: 'middle' },
  textWithIcon: { flexShrink: 1 },
  default: { color: theme.General.textColor },
  primary: { color: theme.General.primaryColor },
  secondary: { color: theme.General.textSecondaryColor },
  disabled: { opacity: 0.6 },
  withIconContainer: { flexDirection: 'row', gap: spacing.Gap['gap-1'] },
  icon: {
    width: typography.Size['text-base'],
    height: typography.Size['text-base'],
  },
}))
