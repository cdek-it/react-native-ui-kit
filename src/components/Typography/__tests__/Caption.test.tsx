import { IconInfoCircle } from '@tabler/icons-react-native'
import { render } from '@testing-library/react-native'

import { Caption, type CaptionProps } from '../Caption'

describe('Caption component tests', () => {
  const snapshotCases = generatePropsCombinations<CaptionProps>({
    color: ['default', 'secondary', 'primary'],
    disabled: [true, false],
    Icon: [undefined, IconInfoCircle],
  })

  test.each(snapshotCases)('color = $color, disabled = $disabled', (props) => {
    const renderedCaption = render(<Caption {...props}>Text</Caption>)

    expect(renderedCaption.toJSON()).toMatchSnapshot()
  })

  test('default props', () => {
    const renderedCaption = render(<Caption>Text</Caption>)

    expect(renderedCaption.toJSON()).toMatchSnapshot()
  })
})
