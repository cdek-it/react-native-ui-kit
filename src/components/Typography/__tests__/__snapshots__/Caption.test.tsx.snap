// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caption component tests color = default, disabled = false 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "rgba(0, 0, 0, 0.8000)",
      },
      false,
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = default, disabled = false 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="rgba(0, 0, 0, 0.8000)"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 3422552064,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.8000)",
        },
        false,
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests color = default, disabled = true 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "rgba(0, 0, 0, 0.8000)",
      },
      {
        "opacity": 0.6,
      },
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = default, disabled = true 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="rgba(0, 0, 0, 0.8000)"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 3422552064,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 3422552064,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.8000)",
        },
        {
          "opacity": 0.6,
        },
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests color = primary, disabled = false 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "#188700",
      },
      false,
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = primary, disabled = false 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="#188700"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 4279797504,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "#188700",
        },
        false,
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests color = primary, disabled = true 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "#188700",
      },
      {
        "opacity": 0.6,
      },
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = primary, disabled = true 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="#188700"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 4279797504,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4279797504,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "#188700",
        },
        {
          "opacity": 0.6,
        },
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests color = secondary, disabled = false 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "rgba(0, 0, 0, 0.6000)",
      },
      false,
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = secondary, disabled = false 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="rgba(0, 0, 0, 0.6000)"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 2566914048,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.6000)",
        },
        false,
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests color = secondary, disabled = true 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "rgba(0, 0, 0, 0.6000)",
      },
      {
        "opacity": 0.6,
      },
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Caption component tests color = secondary, disabled = true 2`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 3.5,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="rgba(0, 0, 0, 0.6000)"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "payload": 2566914048,
          "type": 0,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M12 9h.01"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        d="M11 12h1v4h1"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 2566914048,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      [
        {
          "fontSize": 12.25,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.6000)",
        },
        {
          "opacity": 0.6,
        },
        {
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Caption component tests default props 1`] = `
<Text
  style={
    [
      {
        "fontSize": 12.25,
        "includeFontPadding": false,
        "verticalAlign": "middle",
      },
      {
        "color": "rgba(0, 0, 0, 0.8000)",
      },
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Text
</Text>
`;
